File: 414-divide-two-integers/standard_answer.cc
class Solution {
public:
    /**
     * @param dividend: the dividend
     * @param divisor: the divisor
     * @return: the result
     */
    int divide(int dividend, int divisor) {
        // write your code here
        if (divisor == 0) {
            return dividend >= 0 ? INT_MAX : INT_MIN;
        }
        if (dividend == 0) {
            return 0;
        }
        if (dividend == INT_MIN && divisor == -1) {
            return INT_MAX;
        } 
        bool isNegative = (dividend < 0 && divisor > 0) || 
                          (dividend > 0 && divisor < 0);
        long a = abs(long(dividend));
        long b = abs(long(divisor));
        int result = 0;
        while (a >= b) {
            int shift = 0;
            while (a >= (b << shift)) {
                shift++;
            }
            a -= b << (shift - 1);
            result += 1 << (shift - 1);
        }
        return isNegative ? -result : result;
    }
        
};

File: 647-find-all-anagrams-in-a-string/standard_answer.java
public class Solution {
    /**
     * @param s: a string
     * @param p: a string
     * @return: a list of index
     */
    public List<Integer> findAnagrams(String s, String p) {
        // write your code here
        int[] hash = new int[256];
        char[] sc = s.toCharArray();
        char[] pc = p.toCharArray();
        List<Integer> result = new ArrayList<>();
        if (pc.length > sc.length) {
            return result;
        }
        for (int i = 0; i < pc.length; i++) {
            hash[pc[i]]--;
            hash[sc[i]]++;
        }
        int addSum = 0;
        for (int item : hash) {
            addSum += Math.abs(item);
        }
        if (addSum == 0) {
            result.add(0);
        }
        for (int i = pc.length; i < sc.length; i++) {
            int r = sc[i];
            int l = sc[i - pc.length];
            addSum = addSum - Math.abs(hash[r]) - Math.abs(hash[l]); // don't know why
            hash[r]++;
            hash[l]--;
            addSum = addSum + Math.abs(hash[r]) + Math.abs(hash[l]);
            if (addSum == 0) {
                result.add(i - pc.length + 1);
            }
        }
        return result;
    }
}
